akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StashSpec.scala

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+package akka.actor.typed
+package scaladsl
+
+import akka.event.LoggingAdapter
+import akka.testkit.typed.TestKit
+import akka.testkit.typed.scaladsl.TestProbe
+
+object StashSpec {
+  sealed trait Command
+  final case class Msg(s: String) extends Command
+  final case class Unstashed(cmd: Command) extends Command
+
+  case object Stash extends Command
+  case object UnstashAll extends Command
+  case object Unstash extends Command
+  final case class GetProcessed(replyTo: ActorRef[Vector[String]]) extends Command
+
+  // FIXME replace when we get the logging in place, #23326
+  def log(ctx: ActorContext[_]): LoggingAdapter = ctx.system.log
+
+  def active(processed: Vector[String]): Behavior[Command] =
+    Behaviors.immutable { (ctx, cmd) ⇒
+      cmd match {
+        case msg: Msg ⇒
+          active(processed :+ msg.s)
+        case GetProcessed(replyTo) ⇒
+          replyTo ! processed
+          Behaviors.same
+        case Stash ⇒
+          stashing(ImmutableStashBuffer.empty(10), processed)
+        case UnstashAll ⇒
+          Behaviors.unhandled
+        case Unstash ⇒
+          Behaviors.unhandled
+        case u: Unstashed ⇒
+          throw new IllegalStateException(s"Unexpected $u in active")
+      }
+    }
+
+  def stashing(buffer: ImmutableStashBuffer[Command], processed: Vector[String]): Behavior[Command] =
+    Behaviors.immutable { (ctx, cmd) ⇒
+      cmd match {
+        case msg: Msg ⇒
+          stashing(buffer :+ msg, processed)
+        case GetProcessed(replyTo) ⇒
+          replyTo ! processed
+          Behaviors.same
+        case UnstashAll ⇒
+          buffer.unstashAll(ctx, active(processed))
+        case Unstash ⇒
+          log(ctx).debug(s"Unstash ${buffer.size}")
+          if (buffer.isEmpty)
+            active(processed)
+          else {
+            ctx.self ! Unstash // continue unstashing until buffer is empty
+            val numberOfMessages = 2
+            log(ctx).debug(s"Unstash $numberOfMessages of ${buffer.size}, starting with ${buffer.head}")
+            buffer.unstash(ctx, unstashing(buffer.drop(numberOfMessages), processed), numberOfMessages, Unstashed)
+          }
+        case Stash ⇒
+          Behaviors.unhandled
+        case u: Unstashed ⇒
+          throw new IllegalStateException(s"Unexpected $u in stashing")
+      }
+    }
+
+  def unstashing(buffer: ImmutableStashBuffer[Command], processed: Vector[String]): Behavior[Command] =
```

Comment:
I was thinking that it would be good to be able to support unstashing a few messages at a time, to not block the actor in processing too many messages. As you can see it became rather complicated because one would have to know if it's an unstashed message or a new incoming message. The new messages must be added to the stash to get the right order.

Reply:
i think it is a useful feature but we also shouldn't encourage stashes that are too large